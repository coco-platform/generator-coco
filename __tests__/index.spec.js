/**
 * @description - coco generator
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
const path = require('path');
const fs = require('fs');
const renderer = require('yeoman-test');
const assert = require('yeoman-assert');

describe('coco:webpack', () => {
  it('should generate webpack configuration with typescript', () => {
    const webpack = path.resolve(__dirname, '../webpack');
    const prompts = {
      typescript: true,
      cssmodules: false,
      definition: 'bootcdn.stable.yml',
      entry: './src/main.js',
    };

    return renderer
      .run(webpack)
      .withPrompts(prompts)
      .then(() => {
        assert.file(['webpack.development.js', 'webpack.production.js']);

        const development = fs.readFileSync('webpack.development.js', 'utf8');
        const production = fs.readFileSync('webpack.production.js', 'utf8');

        expect(development).toMatchSnapshot();
        expect(production).toMatchSnapshot();
      });
  });

  it('should generate webpack configuration without typescript', () => {
    const webpack = path.resolve(__dirname, '../webpack');
    const prompts = {
      typescript: false,
      cssmodules: false,
      definition: 'bootcdn.stable.yml',
      entry: './src/main.js',
    };

    return renderer
      .run(webpack)
      .withPrompts(prompts)
      .then(() => {
        assert.file(['webpack.development.js', 'webpack.production.js']);

        const development = fs.readFileSync('webpack.development.js', 'utf8');
        const production = fs.readFileSync('webpack.production.js', 'utf8');

        expect(development).toMatchSnapshot();
        expect(production).toMatchSnapshot();
      });
  });
});
