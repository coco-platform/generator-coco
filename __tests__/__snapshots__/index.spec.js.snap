// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coco:webpack should generate projects with preset webpack 1`] = `
"/**
 * @description - webpack development configuration
 * @author - huang.jian
 */

// packages
const path = require('path');
const webpack = require('webpack/index');
const CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin');
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const InjectExternalPlugin = require('@coco-platform/webpack-plugin-inject-external');
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');

module.exports ={
  mode: 'development',
  target: 'web',
  entry: {
    main: path.resolve(process.cwd(), './src/main.js'),
  },
  output: {
    path: path.resolve(process.cwd(), 'dist', 'client'),
    publicPath: '/',
    filename: 'static/script/[name].bundle.js',
    chunkFilename: 'static/script/[id]_[name].chunk.js',
    crossOriginLoading: 'anonymous',
  },
  resolve: {
    extensions: [
      '.js',
      '.jsx',
      '.mjs',
      '.json',
      '.ts',
      '.tsx',
    ],
    alias: {},
  },
  module: {
    noParse: [/\\\\.min\\\\.js/],
    rules: [
      {
        test: /\\\\.(js|jsx|mjs|mjsx|ts|tsx)$/,
        exclude: /node_modules/,
        include: path.resolve(process.cwd(), 'src'),
        use: [
          {
            loader: require.resolve('babel-loader'),
            options: {
              cacheDirectory: true,
            },
          },
        ],
      },
      {
        test: /\\\\.p?css$/,
        exclude: /node_modules/,
        use: [
          require.resolve('style-loader'),
          {
            loader: require.resolve('css-loader'),
            options: {
              root: path.resolve(process.cwd(), 'src'),
              modules: false,
              importLoaders: 1,
              localIdentName: '[name]__[local]__[hash:base64:5]',
            },
          },
          {
            loader: require.resolve('postcss-loader'),
            options: {
              ident: 'postcss',
            },
          },
        ],
      },
      {
        test: /\\\\.css$/,
        exclude: [
          path.resolve(process.cwd(), 'src'),
          path.resolve(process.cwd(), 'public'),
        ],
        use: [
          { loader: require.resolve('style-loader') },
          { loader: require.resolve('css-loader') },
        ],
      },
      {
        test: /\\\\.(png|jpe?g|gif|bmp|mp3|woff|woff2|ttf|eot|svg)(\\\\?.*)?$/,
        use: [
          {
            loader: require.resolve('file-loader'),
            options: {
              name: 'static/media/[name].[hash:8].[ext]',
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new webpack.WatchIgnorePlugin([/\\\\.js$/, /\\\\.d\\\\.ts$/]),
    new ForkTsCheckerWebpackPlugin(),
    new CaseSensitivePathsPlugin(),
    new webpack.ContextReplacementPlugin(/moment\\\\/locale$/, /zh-cn/),
    new HtmlWebpackPlugin({
      inject: 'body',
      template: path.resolve(process.cwd(), 'public', 'index.html'),
      favicon: path.join(process.cwd(), 'public', 'favicon.ico'),
    }),
    
    new InjectExternalPlugin({
      env: 'development',
      definition: bootcdn.stable.yml,
    }),
    
    new BundleAnalyzerPlugin({
      analyzerMode: 'server',
      openAnalyzer: false,
      reportFilename: '../analyzer/index.html',
    }),
  ],
  optimization: {
    runtimeChunk: {
      name: 'manifest',
    },
  },
  devtool: 'cheap-module-source-map',
  node: {
    dgram: 'empty',
    fs: 'empty',
    net: 'empty',
    crypto: 'empty',
    tls: 'empty',
    child_process: 'empty',
  },
};
"
`;

exports[`coco:webpack should generate projects with preset webpack 2`] = `
"/**
 * @description - declare webpack production configuration
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
const path = require('path');
const webpack = require('webpack');
const CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const ProgressBarPlugin = require('progress-bar-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const CompressionPlugin = require('compression-webpack-plugin');
const InjectExternalPlugin = require('@coco-platform/webpack-plugin-inject-external');
const HtmlMinifyPlugin = require('@coco-platform/webpack-plugin-html-minify');
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');

module.exports = {
  mode: 'production',
  target: 'web',
  entry: {
    main: path.resolve(process.cwd(), './src/main.js'),
  },
  output: {
    path: path.resolve(process.cwd(), 'dist', 'client'),
    publicPath: '/',
    filename: 'static/script/[name].[chunkhash:8].js',
    chunkFilename: 'static/script/[id]_[name]_[chunkhash:8].chunk.js',
    crossOriginLoading: 'anonymous',
  },
  resolve: {
    extensions: [
      '.js',
      '.jsx',
      '.mjs',
      '.json',
      '.ts',
      '.tsx',
      '.web.js',
      '.web.jsx',
    ],
    alias: {},
  },
  module: {
    rules: [
      {
        test: /\\\\.(js|jsx|mjs|mjsx|ts|tsx)$/,
        include: path.resolve(process.cwd(), 'src'),
        use: [
          {
            loader: require.resolve('babel-loader'),
            options: {
              cacheDirectory: true,
            },
          },
        ],
      },
      {
        test: /\\\\.p?css$/,
        exclude: /node_modules/,
        use: [
          MiniCssExtractPlugin.loader,
          {
            loader: require.resolve('css-loader'),
            options: {
              root: path.resolve(process.cwd(), 'src'),
              modules: false,
              importLoaders: 1,
              localIdentName: '[name]__[local]__[hash:base64:5]',
            },
          },
          {
            loader: require.resolve('postcss-loader'),
            options: {
              ident: 'postcss',
            },
          },
        ],
      },
      {
        test: /\\\\.css$/,
        exclude: [
          path.resolve(process.cwd(), 'src'),
          path.resolve(process.cwd(), 'public'),
        ],
        use: [
          MiniCssExtractPlugin.loader,
          { loader: require.resolve('css-loader') },
        ],
      },
      {
        test: /\\\\.(png|jpe?g|gif|bmp|mp3|woff|woff2|ttf|eot|svg)(\\\\?.*)?$/,
        use: [
          {
            loader: require.resolve('file-loader'),
            options: {
              name: 'static/media/[name].[hash:8].[ext]',
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new ProgressBarPlugin(),
    new CaseSensitivePathsPlugin(),
    new MiniCssExtractPlugin({
      filename: 'static/stylesheet/[name].[hash].css',
      chunkFilename: 'static/stylesheet/[id].[hash].chunk.css',
    }),
    new webpack.ContextReplacementPlugin(/moment\\\\/locale$/, /zh-cn/),
    new HtmlWebpackPlugin({
      inject: 'body',
      template: path.resolve(process.cwd(), 'public', 'index.html'),
      favicon: path.join(process.cwd(), 'public', 'favicon.ico'),
    }),
    new CompressionPlugin({
      test: /\\\\.(js|css|html)$/,
      threshold: 1024,
      minRatio: 0.85,
    }),
    
    new InjectExternalPlugin({
      env: 'development',
      definition: bootcdn.stable.yml,
    }),
    
    new HtmlMinifyPlugin(),
    new BundleAnalyzerPlugin({
      analyzerMode: 'static',
      reportFilename: '../analyzer/index.html',
      openAnalyzer: false,
    }),
  ],
  optimization: {
    runtimeChunk: {
      name: 'manifest',
    },
    minimize: true,
    minimizer: [
      new UglifyJsPlugin({
        cache: true,
        parallel: true,
        sourceMap: true, // set to true if you want JS source maps
      }),
      new OptimizeCSSAssetsPlugin({}),
    ],
  },
  devtool: 'source-map',
  node: {
    dgram: 'empty',
    fs: 'empty',
    net: 'empty',
    crypto: 'empty',
    tls: 'empty',
    child_process: 'empty',
  },
};
"
`;
